eip-5564.md

```diff
@@ -8,174 +8,170 @@ status: Draft
-        // Retrieve public key coordinates of target
-        (uint256 PublicKeyX, uint256 PublicKeyY)  = stealthKeys(target);
-        //  s*G = S
-        pubkeyFromSecret = ecMul(secret, G);
-        //  s*P = q
-        sharedSecret = ecMul(secret, Publickey);
-        // hash(sharedSecret)
-        hashedSharedSecret = keccak256(sharedSecret);
-        // hash value to public key
-        pubkeyFromHashedSecret = ecMul(hashedSharedSecret, G);
-        // derive new public key
-        stealthPubkeyRecipient = ecAdd(Publickey, pubkeyFromHashedSecret);
-        // generate stealth address
-        stealthAddressRecipient = address(stealthPubkeyRecipient);
-        // return public key coordinates and stealthAddress
-        return (pubkeyFromSecret, stealthAddressRecipient);
-    }
-```
-
-
----
-
-#### For transfering:
-
-Recipient Keypair --> $(p,P) | P = p * G$
-
-
-##### Recipient:
-```solidity
-(uint256 PublicKeyX, uint256 PublicKeyY)  = stealthKeys(target);
-```
-NOTE: `pubkeyRecipient` is stored in the `keys` mapping and is retrieved from there.
-
----
-
-##### Sender:
-
-Sender senderKeypair  --> $(s,S) | S = s * G $
-```solidity
-bytes pubkeyFromSecret = ecMul(s, G);
-```
-NOTE: The parameter `s` represents a sender-generated secret and MUST NOT equal a user's private key.
-
-$sharedSecret    = s * P$
-
-```solidity
-bytes sharedSecret = ecMul(s, pubkeyRecipient);
-```
-
-
-stealthAddress  = $pubtoaddr(P + (G * keccak(sharedSecret)))$
-```solidity
-bytes32 hashedSharedSecret = keccak256(sharedSecret);
-bytes pubkeyFromHashedSecret = ecMul(hashedSharedSecret, G);
-bytes stealthPubkeyRecipient = ecAdd(pubkeyRecipient, pubkeyFromHashedSecret);
-address stealthAddressRecipient = address(stealthPubkeyRecipient);
-```
-Transfer MUST emit a `Announcement` Event containing `S` and the `stealthAddressRecipient`.
 
+/// @notice Interface for generating stealth addresses for keys on a given curve.
+/// @dev The Generator contract SHOULD have a method called decodeKeys that returns the recipient's
+/// public keys as the correct types. This is not included below since the return parameters may
+/// differ for each Generator.
+interface IERC5564Generator {
+  /// @notice Given a `registrant` and `entropy`, returns all relevant data to compute a stealth
+  /// address.
+  /// @dev MUST return all zeroes if the registrant has not registered keys for this generator.
+  /// @dev Intended to be used off-chain only to prevent exposing secrets on-chain.
+  /// @dev Consider running this against a local node, or using an off-chain library with the same
+  /// logic, instead of via an `eth_call` to a public RPC provider to avoid leaking secrets.
+  /// @dev The `entropy` is passed as `bytes memory` to support various keys formats and encryption
+  /// schemes. Consult the generator's documentation for details on how to format this parameter.
+  function generateStealthAddress(address registrant, bytes memory entropy)
```

Comment:
I'm concerned about the pattern of including a generateStealthAddress method on the Generator (or "Lens") contract. I see why it's convenient, but it's just inviting implementors to make the mistake of sending secrets to node operators and thus losing the privacy benefits.

A way to frame it is "bad UX" for the standard. It's like having a big red button in your UI with a message that says "Don't press this button." Someone is gonna press it for sure 🙈

Perhaps having the `decodeKeys` method on each Generator (or we could call them "Lens" contracts, since they won't actually generate the address) is sufficient. It can't be generalized, because the return types will change, but perhaps the pattern is sufficient? Solidity Generics might help here too, whenever they are implemented.

Another possibility would be some on-chain anchor to a reference implementation that *can't* be called externally, to prevent this sort of mistake. Could be something as simple as requiring a content hash to code of a reference implementation stored somewhere like IPFS. Or perhaps requiring an `internal` method that implements it (this would be "dead code" only useful for off-chain observation). Just spitballing ideas.

I'd love to hear what everyone else thinks about this. I could definitely be convinced that including the callable reference implementation is worth the risk it will be misused, but right now I lean toward not.

Reply:
I think, the current ens contract have a similar method for generating the secret commitment for registering. It's also a function that should only be executed locally. 
I guess the only ones that could fall for the described mistake are rushing devs who call contracts without a frontend. Users, on the other side, may just trust their frontend to either call it locally or have the same logic implemented.
Anyway, I agree with the red button analogy, even if the red button is only visible to "advanced" users.
Additionally, as you say, they "only" lose their privacy (e.g. for ens it would lead to unveiling the desired ens name), which is acceptable, assuming that 99% will implement it the right way.
I think we should include that function. It may also help users better understanding the mechanism.